atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftQueryWeighers.java

```diff
@@ -0,0 +1,196 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.cassandra.qos;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import org.apache.cassandra.thrift.ColumnOrSuperColumn;
+import org.apache.cassandra.thrift.CqlResult;
+import org.apache.cassandra.thrift.KeyRange;
+import org.apache.cassandra.thrift.KeySlice;
+import org.apache.cassandra.thrift.Mutation;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Suppliers;
+import com.palantir.atlasdb.keyvalue.cassandra.CassandraClient;
+import com.palantir.atlasdb.qos.ImmutableQueryWeight;
+import com.palantir.atlasdb.qos.QosClient;
+import com.palantir.atlasdb.qos.QueryWeight;
+
+public final class ThriftQueryWeighers {
+
+    private static final Logger log = LoggerFactory.getLogger(CassandraClient.class);
+
+    @VisibleForTesting
+    private static final int ESTIMATED_NUM_BYTES_PER_ROW = 100;
+    static final QueryWeight DEFAULT_ESTIMATED_WEIGHT = ImmutableQueryWeight.builder()
+            .numBytes(ESTIMATED_NUM_BYTES_PER_ROW)
+            .numDistinctRows(1)
+            .timeTakenNanos(TimeUnit.MILLISECONDS.toNanos(2))
+            .build();
+
+    static final QueryWeight ZERO_ESTIMATED_WEIGHT = ImmutableQueryWeight.builder()
+            .numBytes(0)
+            .numDistinctRows(1)
+            .timeTakenNanos(TimeUnit.MILLISECONDS.toNanos(1))
+            .build();
+
+    private ThriftQueryWeighers() { }
+
+    static QosClient.QueryWeigher<Map<ByteBuffer, List<ColumnOrSuperColumn>>> multigetSlice(List<ByteBuffer> keys,
+            boolean zeroEstimate) {
+        return zeroEstimate
+                ? readWeigherWithZeroEstimate(ThriftObjectSizeUtils::getApproximateSizeOfColsByKey, Map::size,
+                keys.size())
+                : readWeigher(ThriftObjectSizeUtils::getApproximateSizeOfColsByKey, Map::size, keys.size());
+    }
+
+    static QosClient.QueryWeigher<List<KeySlice>> getRangeSlices(KeyRange keyRange, boolean zeroEstimate) {
+        return zeroEstimate
+                ? readWeigherWithZeroEstimate(ThriftObjectSizeUtils::getApproximateSizeOfKeySlices, List::size,
+                keyRange.count)
+                : readWeigher(ThriftObjectSizeUtils::getApproximateSizeOfKeySlices, List::size, keyRange.count);
+    }
+
+    static QosClient.QueryWeigher<ColumnOrSuperColumn> get(boolean zeroEstimate) {
+        return zeroEstimate
+                ? readWeigherWithZeroEstimate(ThriftObjectSizeUtils::getColumnOrSuperColumnSize, ignored -> 1, 1)
+                : readWeigher(ThriftObjectSizeUtils::getColumnOrSuperColumnSize, ignored -> 1, 1);
+    }
+
+    static final QosClient.QueryWeigher<CqlResult> EXECUTE_CQL3_QUERY =
```

Comment:
`EXECUTE_CQL3_QUERY` is a misleading name for the weigher. Can we estimate the number of rows from the `CqlResult`, each CqlRow (partition) has a list of columns (each of which has a ts) - is `CqlResult.getRows().stream().mapToInt(row -> row.getColumns.size()).sum()` the estimate here?

Reply:
a `CqlRow` is not the same as a cassandra row. e.g., if i `select (key, column1, column2)` i will get 1 `CqlRow` per column, but there's a difference between having 1000 columns in a single row vs 1000 rows with a single column